File: server-api/data-access/mongo-db/api/logout.js
import { Types } from 'mongoose';

const logout = ({ User }) => async ({ userId: _userId, refreshToken, everywhere }) => {
  if (_userId) {
    const userId = new Types.ObjectId(_userId);
    const user = await User.findOne({ _id: userId });
    const logoutSingle = user && !everywhere && refreshToken;
    const logoutEverywhere = user && everywhere;
    if (logoutSingle || logoutEverywhere) {
      const operations = logoutSingle
        ? { $pull: { tokens: refreshToken } }
        : { $set: { tokens: [] } };
      await User.findOneAndUpdate(
        { _id: userId },
        operations,
      );
    }
  }
  return true; // success
};

export default logout;

File: server-rust/data-access/src/api/logout.rs
use bson::{doc, oid::ObjectId, Document};
use mongodb::Database;

use crate::{error::Error, model::User};

pub async fn logout(
    database: Database,
    user_id: ObjectId,
    refresh_token: String,
    everywhere: bool,
) -> Result<(), Error> {
    let users = database.collection::<User>("user");

    users
        .find_one(doc! { "id" : user_id}, None)
        .await?
        .ok_or(Error::UserNotFound)?;

    let operation: Document;
    if everywhere {
        operation = doc! { "$set" : { "tokens" : []}};
    } else {
        operation = doc! { "$pull" : { "tokens" : refresh_token}};
    }

    let _ = users
        .find_one_and_update(doc! { "id" : user_id}, operation, None)
        .await?;

    Ok(())
}
